chain/indexer/processor/state.go

```diff
@@ -0,0 +1,326 @@
+	// Implementation of this interface are responsible for abort when their context is canceled.
+	ProcessTipSets(ctx context.Context, current *types.TipSet, executed *types.TipSet) (model.Persistable, *visormodel.ProcessingReport, error)
+}
+
+type ActorProcessor interface {
+	// ProcessActors processes a set of actors. If error is non-nil then the processor encountered a fatal error.
+	// Any data returned must be accompanied by a processing report.
+	// Implementation of this interface are responsible for abort when their context is canceled.
+	ProcessActors(ctx context.Context, current *types.TipSet, executed *types.TipSet, actors tasks.ActorStateChangeDiff) (model.Persistable, *visormodel.ProcessingReport, error)
+}
+
+type ReportProcessor interface {
+	// ProcessTipSet processes a tipset. If error is non-nil then the processor encountered a fatal error.
+	// Implementation of this interface are responsible for abort when their context is canceled.
+	ProcessTipSet(ctx context.Context, current *types.TipSet) (visormodel.ProcessingReportList, error)
+}
+
+var log = logging.Logger("lily/index/processor")
+
+type StateProcessor struct {
+	builtinProcessors map[string]ReportProcessor
+	tipsetProcessors  map[string]TipSetProcessor
+	tipsetsProcessors map[string]TipSetsProcessor
+	actorProcessors   map[string]ActorProcessor
+
+	// api used by tasks
+	api tasks.DataSource
+
+	//pwg is a wait group used internally to signal processors completion
+	pwg sync.WaitGroup
+
+	// name of the processor
+	name string
+}
+
+// A Result is either some data to persist or an error which indicates that the task did not complete. Partial
+// completions are possibly provided the Data contains a persistable log of the results.
+type Result struct {
+	Task        string
+	Error       error
+	Report      visormodel.ProcessingReportList
+	Data        model.Persistable
+	StartedAt   time.Time
+	CompletedAt time.Time
+}
+
+// State executes its configured processors in parallel, processing the state in `current` and `executed. The return channel
+// emits results of the state extraction closing when processing is completed. It is the responsibility of the processors
+// to abort if its context is canceled.
+// A list of all tasks executing is returned.
+func (sp *StateProcessor) State(ctx context.Context, current, executed *types.TipSet) (chan *Result, []string) {
+	ctx, span := otel.Tracer("").Start(ctx, "StateProcessor.State")
+
+	num := len(sp.tipsetProcessors) + len(sp.actorProcessors) + len(sp.tipsetsProcessors) + len(sp.builtinProcessors)
+	results := make(chan *Result, num)
+	taskNames := make([]string, 0, num)
+
+	taskNames = append(taskNames, sp.startReport(ctx, current, results)...)
+	taskNames = append(taskNames, sp.startTipSet(ctx, current, results)...)
+	taskNames = append(taskNames, sp.startTipSets(ctx, current, executed, results)...)
+	taskNames = append(taskNames, sp.startActor(ctx, current, executed, results)...)
```

Comment:
I'd like a generic method that can be used to start any type of processor, but I couldn't figure out how to do that here. Ideas?

Reply:
Maybe you need to pass these functions a type with the following interface:

```Go
type TaskData interface {
    Current() *types.TipSet
    Executed() *types.TipSet
    Actors()  tasks.ActorStateChangeDiff
}
```
Each processor could then have the same signature:

```Go
Process(ctx context.Context, td TaskData) (model.Persistable, *visormodel.ProcessingReport, error)
```

Reply:
oh I like this. Then the `Actors()` method implementation would memoize the result of ActorStateChanges and let all processors to call the method internally, similar to what we do now for ExectuedMessage and such. 

This will simplify the StateProcessor down to two processor interfaces:
```go
type Data interface {
	Current() *types.TipSet
	Executed() *types.TipSet
	Actors(ctx context.Context) (tasks.ActorStateChangeDiff, error)
}

type ModelProcessor interface {
	Process(ctx context.Context, d Data) (model.Persistable, *visormodel.ProcessingReport, error)
}

type ReportProcessor interface {
	Process(ctx context.Context, current *types.TipSet) (visormodel.ProcessingReportList, error)
}
```

Given the amount of code churn implementing this now will cause I'd prefer to do this as a follow on once we settle on design here, thoughts?